CMakeLists.txt

```diff
@@ -0,0 +1,31 @@
+cmake_minimum_required(VERSION 3.1.3 FATAL_ERROR)
+project(hidapi LANGUAGES C)
+
+if(NOT DEFINED CMAKE_BUILD_TYPE OR CMAKE_BUILD_TYPE STREQUAL "")
+    set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Debug;Release;MinSizeRel;RelWithDebInfo" FORCE)
+endif()
+set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
+
+if(APPLE)
+    option(CMAKE_FRAMEWORK "Build macOS/iOS Framework version of the library" OFF)
+elseif(NOT WIN32)
+    if(CMAKE_SYSTEM_NAME MATCHES "Linux")
+        option(HIDAPI_WITH_HIDRAW "Build HIDRAW-based implementation of HIDAPI" ON)
+        option(HIDAPI_WITH_LIBUSB "Build LIBUSB-based implementation of HIDAPI" ON)
+    endif()
+endif()
+
+option(BUILD_SHARED_LIBS "When set: build shared version of the libraries, otherwise - static" ON)
+
+set(HIDAPI_INSTALL_TARGETS ON)
```

Comment:
unfortunately you don't know the design that is hidden behind it

in short - the root CMakeLists.txt is not designed to be a subdirectory of a larger project `src/CMakeLists.txt` is designed for that
I'll surely explain that in the documentation/usage examples as soon as I can get my hands to it

Reply:
>  the root CMakeLists.txt is not designed to be a subdirectory of a larger project

It should be. That's how CMake libraries should be structured. The fact they can be installed system-wide is an extra thing, not the default.

You cut out users that want to statically link without polluting their system cache, or those who are cross compiling, etc.

Reply:
> It should be. That's how CMake libraries should be structured.

I don't suppose that statement is based on large number of different real-life CMake-based projects/libraries.

---
I had a pleasure integrating maybe a dozen CMake-based libraries into a larger CMake-based solution/build tree (as a submodule). And none of those projects could be integrated by simple `add_subdirectory(<subproject root>)`, and I don't believe that is because those projects or their build systems are badly implemented, not at all.
Although, a solution/build tree I mentioned above has lots of demands, some of them: it has to compile/deploy on at least 4 platforms: Windows, macOS, Linux, Android, including host x86_64 builds and ARM64 cross-compilations. All these and other requirements automatically applies to all dependencies.

Per my observations, each projects's build system consist of of two logical parts: _build infrastructure_ and _build definitions_. (It is not always possible to tell exactly where the infrastructure ends, and the definitions begins, but still.)
CMake itself is an infrastructure. It is designed to suit as much different possible needs as possible. That also means that it will never work of-the-box for everyone. CMake has many [variables](https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html) and [properties](https://cmake.org/cmake/help/latest/manual/cmake-properties.7.html) to configure the build system, but it doesn't mean, that any combination of those values/properties can by applied to any project successfully. So each project (its author(s)) choose a subset of its variables/properties, maybe its values, into an actual CMake options that are intended to be configurable/changed. That all together makes a build infrastructure, that answers to question "how?": how to build a target, which compiler options, where to install, what kind of the binary should be, etc.
At this point we didn't even start developing our project, we don't have build definitions yet, we don't have our "what". Q: _what_ do we want to? - Make a library. So we add a build definition (a target): `add_library(...)`. There are certain properties of each target, that are absolutely required to successfully build a target, e.g.: macOS backend of hidapi has to be linked against `IOKit framework`, otherwise of won't build at all. That is essential, that has to be part of build definition. It cannot be optional or configurable.
On the other hand, we can build our source with `-O2` or `-g -O0` compiler options, but that doesn't change the essence/intention of the project. Maybe performance, at most, maybe convenience for the developer (e.g. ability to debug), and that is a part of the project's infrastructure.

My point is that root CMakeLists.txt usually contains a lot of infrastructure for convenience of the project's developer. Because project's developer keeps its focus on the ability to integrate the project into a package manager, install into a system (and make the library available for other projects), deploy it to production, ability to develop it fast and efficient.
hidapi is not an exception here. All the options are designed for _maximum convenience of the user/developer_, who tries to build hidapi as a _standalone project/library_. Install project or not - that is a question to the user: call/build `install` target or not, it shouldn't be a question to the build system as "add `install` target or not?".
I tried to cover all possible cases I could thing of. Most CMake-based projects I've seen so far, are actually doing the same in one degree or another. Without a good infrastructure, a usage of the project's build system is not impossible, but less convenient/harder for sure.

On the other hand, when someone wan'ts to use a hidapi as a subdirectory of the existing CMake build tree, it likely means they _already have its own build infrastructure_. CMake options that _make perfect sense for hidapi_ might have absolutely _no sense to the other project_ that want's to use hidapi (e.g. as exactly a single static library). So the only thing other project needs from hidapi - its _build definitions_, and minimum dependencies.
By current design, root CMakeLists.txt of hidapi contains an _infrastructure_. `src/CMakeLists.txt` and below - that's definitions. As an author of the build system, I've tried to design it that way, and only because I've thought of all that even before I've started the development.
Unfortunately not all of the infrastructure can be removed from definitions (at least, because some of the infrastructure is actually what makes definitions possible), so all of the infrastructure parts that are left are designed to be configurable as much as possible, to cover the majority of possible usages.

Reply:
You can detect whether you're the main project or included as a subproject as follows:
```cmake
cmake_minimum_required(VERSION 3.0)
if(PROJECT_SOURCE_DIR)
    message(STATUS "Included as a subproject")
    set(HIDAPI_MAIN_PROJECT OFF)
else()
    message(STATUS "I'm the main project")
    set(HIDAPI_MAIN_PROJECT ON)
endif()
project(libusb)
```

That way, there's no need for two `CMakeLists.txt` files.

Reply:
The two `CMakeLists.txt` files are still there, just because the defaults/behavior will be a bit different, when built as a sub_directory and as a standalone project, but yes, this idea is already implemented: https://github.com/libusb/hidapi/commit/81291b205b52013019c2498620b6d27f2ee3bcc9

And technically, checking for `PROJECT_SOURCE_DIR` isn't the same as checking for `CMAKE_/CURRENT_/SOURCE_DIR`, and isn't strictly correct. No one prohibits you to do something like:
```cmake
cmake_minimum_required(VERSION 3.0)

add_subdirectory(hidapi)

project(my_awsome_project)

#...
```